tests/ui/async-await/dyn/works.rs

```diff
@@ -0,0 +1,30 @@
+//@ aux-build:block-on.rs
+//@ edition: 2021
+//@ run-pass
+//@ check-run-results
+
+#![feature(async_fn_in_dyn_trait)]
+//~^ WARN the feature `async_fn_in_dyn_trait` is incomplete
+
+extern crate block_on;
+
+use std::future::Future;
+
+trait AsyncTrait {
+    async fn async_dispatch(&self);
+}
+
+impl AsyncTrait for &'static str {
+    fn async_dispatch(&self) -> impl Future<Output = ()> {
+        Box::pin(async move {
```

Comment:
I'd also expect other pointer-sized `Future` implementations to work here, [e.g. `Stream::next`](https://docs.rs/futures-util/0.3.31/src/futures_util/stream/stream/next.rs.html#22).

Reply:
I'll add a test for `Pin<&mut Fut>` but I probably will not add more since I just want to get this landed :>

Reply:
@cramertj:

> I'd also expect other pointer-sized Future implementations to work here, e.g. Stream::next.

Not since https://github.com/rust-lang/rust/pull/133226, which made the `PointerLike` trait no longer automatically be implemented for things based on their layout. Instead, we'll need to eventually add a `derive(PointerLike)` for APIs that want to promise in their API that their futures are pointer-like (and thus coercible to `dyn* Future`).